File: src/main/java/com/mattchowning/file_read_write/server/ServerUtil.java
package com.mattchowning.file_read_write.server;

import com.mattchowning.file_read_write.server.model.HttpError;
import com.mattchowning.file_read_write.server.model.OAuthToken;

import org.jetbrains.annotations.NotNull;

import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;
import java.util.Base64;

import io.netty.buffer.Unpooled;
import io.netty.channel.ChannelHandlerContext;
import io.netty.handler.codec.http.*;
import io.netty.util.CharsetUtil;

import static com.mattchowning.file_read_write.SharedConstants.GSON;
import static com.mattchowning.file_read_write.SharedConstants.RESPONSE_CHARSET;

public class ServerUtil {

    public OAuthToken getOAuthToken(String encodedAuthorizationHeader) {
        String tokenType = getTokenType(encodedAuthorizationHeader);
        String token = getAccessToken(encodedAuthorizationHeader);
        return new OAuthToken(token, tokenType);
    }

    @NotNull
    private static String getAccessToken(@NotNull String authorizationHeader) {
        String[] authHeaderArray = authorizationHeader.split("\\s");
        if (authHeaderArray.length < 2) {
            return "";
        } else {
            String encodedToken = authHeaderArray[1];
            byte[] decodedTokenBytes = Base64.getDecoder().decode(encodedToken);
            return new String(decodedTokenBytes, CharsetUtil.UTF_8);
        }
    }

    @NotNull
    private static String getTokenType(@NotNull String encodedAuthorizationHeader) {
        String[] authHeaderElements = encodedAuthorizationHeader.split("\\s");
        return authHeaderElements[0];
    }

    public String getDate() {
        return DateTimeFormatter.RFC_1123_DATE_TIME.format(ZonedDateTime.now());
    }

    public void sendError(ChannelHandlerContext ctx, HttpResponseStatus status) {
        sendError(ctx, status, status.toString());
    }

    public void sendError(ChannelHandlerContext ctx, HttpResponseStatus status, String errorMsg) {
        sendError(ctx, status, errorMsg, null);
    }

    public void sendError(ChannelHandlerContext ctx,
                                 HttpResponseStatus status,
                                 String errorMsg,
                                 String errorDescription) {
        HttpError error = new HttpError(errorMsg, errorDescription);
        String json = GSON.toJson(error);
        FullHttpResponse response = new DefaultFullHttpResponse(HttpVersion.HTTP_1_1,
                                                                status,
                                                                Unpooled.copiedBuffer(json, RESPONSE_CHARSET));
        response.headers().set(HttpHeaderNames.CONTENT_TYPE, "text/plain; charset=UTF-8");
        HttpUtil.setContentLength(response, json.length());
        ctx.writeAndFlush(response);
    }
}

File: src/main/scala/com/mattchowning_scala/file_read_write/server/ServerUtil.scala
package com.mattchowning_scala.file_read_write.server

import com.mattchowning_scala.file_read_write.server.model.{HttpError, OAuthToken}
import com.mattchowning_scala.file_read_write.SharedConstants.GSON
import com.mattchowning_scala.file_read_write.SharedConstants.RESPONSE_CHARSET
import org.jetbrains.annotations.NotNull
import java.time.ZonedDateTime
import java.time.format.DateTimeFormatter
import java.util.Base64
import io.netty.buffer.Unpooled
import io.netty.channel.ChannelHandlerContext
import io.netty.handler.codec.http._
import io.netty.util.CharsetUtil

// TODO make into trait?
class ServerUtil {

  def getOAuthToken(encodedAuthorizationHeader: String): OAuthToken = {
    val tokenType: String = getTokenType(encodedAuthorizationHeader)
    val token: String = getAccessToken(encodedAuthorizationHeader)
    new OAuthToken(token, tokenType)
  }

  def getDate: String =
    DateTimeFormatter.RFC_1123_DATE_TIME.format(ZonedDateTime.now)

  def sendError(ctx: ChannelHandlerContext, status: HttpResponseStatus): Unit =
    sendError(ctx, status, status.toString)

  def sendError(ctx: ChannelHandlerContext,
                status: HttpResponseStatus,
                errorMsg: String): Unit =
    sendError(ctx, status, errorMsg, null)

  def sendError(ctx: ChannelHandlerContext,
                status: HttpResponseStatus,
                errorMsg: String,
                errorDescription: String): Unit = {
    val error: HttpError = new HttpError(errorMsg, errorDescription)
    val json: String = GSON.toJson(error)
    val response: FullHttpResponse = new DefaultFullHttpResponse(HttpVersion.HTTP_1_1,
                                                                 status,
                                                                 Unpooled.copiedBuffer(json,
                                                                                       RESPONSE_CHARSET))
    response.headers.set(HttpHeaderNames.CONTENT_TYPE, "text/plain; charset=UTF-8")
    HttpUtil.setContentLength(response, json.length)
    ctx.writeAndFlush(response)
  }

  private def getAccessToken(@NotNull authorizationHeader: String): String = {
    val authHeaderArray: Array[String] = authorizationHeader.split("\\s")
    if (authHeaderArray.length < 2) {
      ""
    } else {
      val encodedToken: String = authHeaderArray(1)
      val decodedTokenBytes: Array[Byte] = Base64.getDecoder.decode(encodedToken)
      new String(decodedTokenBytes, CharsetUtil.UTF_8)
    }
  }

  private def getTokenType(@NotNull encodedAuthorizationHeader: String): String =
    encodedAuthorizationHeader.split("\\s")(0)
}